libafl/src/inputs/bytessub.rs

```diff
@@ -27,10 +30,199 @@ fn end_index<R>(range: &R, max_len: usize) -> usize
 where
     R: RangeBounds<usize>,
 {
-    match range.end_bound() {
+    let end = match range.end_bound() {
         Bound::Unbounded => max_len,
         Bound::Included(end) => end + 1,
         Bound::Excluded(end) => *end,
+    };
+
+    min(end, max_len)
+}
+
+fn sub_range<R>(outer_range: &Range<usize>, inner_range: R) -> (Bound<usize>, Bound<usize>)
+where
+    R: RangeBounds<usize>,
+{
+    let start =
+        match (outer_range.start_bound(), inner_range.start_bound()) {
+            (Bound::Unbounded, Bound::Unbounded) => Bound::Unbounded,
+            (Bound::Excluded(bound), Bound::Unbounded)
+            | (Bound::Unbounded, Bound::Excluded(bound)) => Bound::Excluded(*bound),
+            (Bound::Included(bound), Bound::Unbounded)
+            | (Bound::Unbounded, Bound::Included(bound)) => Bound::Included(*bound),
+            (Bound::Included(own), Bound::Included(other)) => Bound::Included(own + other),
+            (Bound::Included(own), Bound::Excluded(other))
+            | (Bound::Excluded(own), Bound::Included(other)) => Bound::Excluded(own + other),
+            (Bound::Excluded(own), Bound::Excluded(other)) => Bound::Excluded(own + other + 1),
+        };
+
+    let end = match (outer_range.end_bound(), inner_range.end_bound()) {
+        (Bound::Unbounded, Bound::Unbounded) => Bound::Unbounded,
+        (Bound::Excluded(bound), Bound::Unbounded) => Bound::Excluded(*bound),
+        (Bound::Unbounded, Bound::Excluded(bound)) => Bound::Excluded(outer_range.end - *bound),
+        (Bound::Included(bound), Bound::Unbounded) => Bound::Included(*bound),
+        (Bound::Unbounded, Bound::Included(bound)) => Bound::Included(outer_range.end - *bound),
+        (Bound::Included(own), Bound::Included(other)) => {
+            Bound::Included(min(*own, outer_range.start + other))
+        }
+        (Bound::Included(own), Bound::Excluded(other)) => {
+            Bound::Included(min(*own, outer_range.start + other - 1))
+        }
+        (Bound::Excluded(own), Bound::Included(other)) => {
+            Bound::Included(min(*own - 1, outer_range.start + other))
+        }
+        (Bound::Excluded(own), Bound::Excluded(other)) => {
+            Bound::Excluded(min(*own, outer_range.start + other))
+        }
+    };
+
+    (start, end)
+}
+
+/// An immutable contiguous subslice of a byte slice.
+/// It is mostly useful to cheaply wrap a subslice of a given input.
+///
+/// A mutable version is available: [`BytesSliceMut`].
+#[derive(Debug)]
+pub struct BytesSlice<'a> {
+    /// The (complete) parent input we will work on
+    parent_slice: OwnedSlice<'a, u8>,
+    /// The range inside the parent input we will work on
+    range: Range<usize>,
+}
+
+/// A mutable contiguous subslice of a byte slice.
+/// It is mostly useful to cheaply wrap a subslice of a given input.
+///
+/// An immutable version is available: [`BytesSlice`].
+#[derive(Debug)]
+pub struct BytesSliceMut<'a> {
```

Comment:
`BytesMutSlice` probably, to be consistent

Reply:
I mean, maybe we should change OwnedMutSlice to OwnedSliceMut? I think the name comes from MutPtr from the stdlib but in that case it could make sense to rename, anyway?
For now we should keep it consistent, though

Reply:
Aah we also have `AsSliceMut` so the OwnedMutSlice is actually the one we should change, let's do it for 0.14.